### ISSUE TITLE
I often get the error nats: no heartbeat received

### ISSUE THREAD

**Issue by VuongUranus** [nats-io/nats.go#1622]
### Observed behavior

I often get the error no heartbeat received. Why do I get this error, where have I misconfigured?
When I get this error it seems like my pull consumer has been deleted, I don't know the reason but it seems to be due to exceeding the time configured in the InactiveThreshold attribute configured in the consumer config.

```go
func  ContinuousPolling(cons jetstream.Consumer, f func([]byte)) (jetstream.MessagesContext, error) {
	iter, err := cons.Messages()
	if err != nil {
		log.Error().Msgf("Error when get message from NATS: %v", err)
		return nil, err
	}
	go func() {
		for {
			msg, err := iter.Next()
			if err != nil {
				if errors.Is(err, jetstream.ErrMsgIteratorClosed) || errors.Is(err, jetstream.ErrConsumerDeleted) {
					return
				}

				log.Error().Msgf("Error when consume message. Reson: %v",  err)
				continue
			}
			go f(msg.Data())
			msg.Ack()
		}
	}()

	return iter, nil
}
```

```go
func setDefaultConsumerConfig(subjects []string, consumerName ...string) jetstream.ConsumerConfig {
	name := ""
	if len(consumerName) > 0 {
		name = consumerName[0]
	}

	return jetstream.ConsumerConfig{
		Name:              name,
		DeliverPolicy:     jetstream.DeliverNewPolicy,
		FilterSubjects:    subjects,
		AckPolicy:         jetstream.AckAllPolicy,
		Replicas:          1,
		InactiveThreshold: 1 * time.Minute,
	}
}
```

### Expected behavior

Please explain to me and avoid this situation again.

### Server and client version

nats-server version: 2.10.12

**Comment by withinboredom** [comment]
I believe this happens if you spend too long processing a message in the message loop. You need to take the message off the "queue" and return processing to the library asap (at least that is what I got when trying to debug this issue in my code).

**Comment by Jarema** [comment]
That should not be the case, especially in such a long inactive threshold.
We will however try to replicate it by running a slow workload.

**Comment by withinboredom** [comment]
AFAICT, the issue appears to be that heartbeats are only processed when calling `iter.Next()` so you need to call it again asap.

**Comment by piotrpio** [comment]
This is not correct. The heartbeats are indeed processed in `Next()`, but the timer which triggers the error is only running in the context of each `Next()` execution. So each time you call `iter.Next()` the timer is reset and stopped when the method returns.

**Comment by withinboredom** [comment]
Indeed! Looking at the code, it looks like this error might be coming from networking issues as the heartbeat isn't paused/reset during reconnection:

https://github.com/nats-io/nats.go/blob/8894a274f4f5f00fd602f74cef462184e31d0256/jetstream/pull.go#L608-L630

**Comment by withinboredom** [comment]
I can confirm that the above PR appears to fix the issue -- at least, I haven't seen this message in awhile now.

**Comment by withinboredom** [comment]
@VuongUranus try adding this to your go.mod file and see if you are also not seeing the issue anymore?

```go
replace github.com/nats-io/nats.go => github.com/withinboredom/nats.go v1.34.2-0.20240517073919-f7520e13f38e
```

**Comment by VuongUranus** [comment]
@withinboredom i can't replace `replace github.com/nats-io/nats.go => github.com/withinboredom/nats.go patch-1`

**Comment by VuongUranus** [comment]
@withinboredom the problem still occurs.

**Comment by piotrpio** [comment]
@VuongUranus The issue mentioned by @withinboredom is fixed here: https://github.com/nats-io/nats.go/pull/1643, we'll be merging it soon.

However, I am still not certain that this is indeed your problem. It would be great if you could verify and if you still encounter the issue it would be very helpful if you could answer the questions from this comment: https://github.com/nats-io/nats.go/issues/1622#issuecomment-2068484682

**Comment by VuongUranus** [comment]
My message handling function is in a goroutine so it may not be due to the effects of processing messages for too long.

**Comment by VuongUranus** [comment]
@piotrpio 
I updated to version 1.36.0 but still get the error nats: no heartbeat received

**Comment by svirmi** [comment]
I have also "no heartbeat received" with 1.36.0 version.
Code snippet:
`	for {

		select {

		case <-nats.Ctx.Done():
			return

		default:

			msg, err := consumer.Next()

			if errors.Is(err, jetstream.ErrMsgIteratorClosed) {
				logger.Error("consumer.Next()", slog.String("nats", err.Error()))
				continue

			} else if errors.Is(err, jetstream.ErrNoHeartbeat) {
				logger.Error("consumer.Next() :: jetstream.ErrNoHeartbeat", slog.String("symbol", symbol))
				continue

			} else if err != nil {
				logger.Error("consumer.Next() :: failed to fetch message", slog.String("nats", err.Error()))
				continue
			}

			// msg.Ack()

			err = json.Unmarshal(msg.Data(), &kline)

			if err != nil {
				logger.Error("error in json.Unmarshal(msg.Data(), &kline)", slog.String("nats", err.Error()))
				return
			}

			nats.KLine <- kline
		}
	}`

**Comment by Jarema** [comment]
After you get that error, do you get it once, and then resume normal operation, or there are many consecutive heartbeat errors?
After the error - can you check if the consumer is still there (with consumer info call in client or CLI?)

**Comment by VuongUranus** [comment]
This error occurs repeatedly on a consumer. When this error occurs, I tried searching for the consumer using CLI but could not find it. And I found that when I use `NoAck: true` for stream config and use AckNonePolicy for consumer config, this error is reduced.

**Comment by Jarema** [comment]
This error means that there is some issue with the Consumer, or JetStream.
Sometimes client can recover, sometimes it can't (for example if it is deleted).

Ack should have nothing to do with it. Most probable reason is having consumer with inactivity threshold set, which will go away after given duration of client inactivity.

**Comment by VuongUranus** [comment]
I use Consume function to pull messages so why is the consumer deleted by inactivity threshold?

**Comment by VuongUranus** [comment]
I increased the value of `InactiveThreshold` but the error still occurs as usual. However using `NoAck: true` and `AckPolicy: jetstream.AckNonePolicy` actually reduces errors a lot.

**Comment by VuongUranus** [comment]
@Jarema

**Comment by kaverhovsky** [comment]
@VuongUranus I have encountered this error. It was a consecutive set of `nats: no heartbeat received` just after nats-server have been shutted down. I couldn't grasp why my client wasn't able to reconnect.

The thing was that I missed `ReconnectWait` option in client configuration (not `Jetstream`, but `Core Nats` client). Because of that all reconnection attempts had been exhausted before nats-server have gone up again. That led to client connection close.

Maybe in your case there were some server failure and reconnection attempts exhaustion.

**Comment by yimliuwork** [comment]
Facing the same issue here. I'm using `consumer.Messages()` with version 1.37.0 to pull messages and often get `nats: no heartbeat received` error. What's interesting is that if I drain the subscription and then resubscribe to the same consumer right away, it'll work. Then if I just change the consumer name and start a new subscription, it fails again...

here is the config for the consumer:
```
consConfig := jetstream.ConsumerConfig{
		Name:              queueGroup,
		Durable:           queueGroup,
		Description:       fmt.Sprintf("consumer for queue group: %v", queueGroup),
		AckPolicy:         jetstream.AckExplicitPolicy,
		AckWait:           30 * time.Second,
		MaxDeliver:        10,               
		FilterSubject:     fmt.Sprintf("%v.gamespace.%v", n.mpEnv, gamespace),
		ReplayPolicy:      jetstream.ReplayInstantPolicy,
		MaxAckPending:     5120, 
		HeadersOnly:       false,
		InactiveThreshold: 60 * time.Second,
	}
```

**Comment by Jarema** [comment]
@yimliuwork can you share a snippet of code where you create and then use the consumer?

**Comment by Jarema** [comment]
This is pretty weird. Could you show us how those functions are used?
Maybe it will give us some hint.

**Comment by yimliuwork** [comment]
That's pretty much all the code really. So basically I'm building a grpc server-stream API that distributes messages from a message queue (a consumer) to different clients. The API request payload includes the consumer's name and start sequence. With each request, we create such consumer or we find the consumer if it already exists. Then we create a subscription to that consumer using  `cons.Messages()` and start pulling messages using `subscription.Next()`.

Another issue I just found out from load testing is that this 'no heartbeat' error can happen half way through a stream. Like I have 100 subscriptions to one consumer, and they are pulling messages fine. Then at some point all 100 subscriptions receive a 'no heartbeat' err.

**Comment by brettinternet** [comment]
@Jarema I have a [reproduction here](https://github.com/brettinternet/nats-consumer-account-jetstream-export) alongside another issue from this issue https://github.com/nats-io/nats.go/issues/1703#issuecomment-2309246291

**Comment by ghstahl** [comment]
I also got this error with latest everything.
```json
{"level":"error","config":{"streamName":"webhook_connectors_inbound","consumerName":"consumer_webhook_connectors_inbound_org000_CON001","subject":"","pullMaxMessages":1000,"backoffPolicy":{"exponentialSeconds":1,"maxElapsedSeconds":0,"cycles":8}},"error":"nats: no heartbeat received","caller":"C:/work/mapped/gtm-go/pkg/commoncore/services/nats/consumer/consumer.go:181","time":"2025-03-28T14:01:27-07:00","message":"Error getting message"}
```

Pretty straight forward setup, where I am sending test messages.

**Comment by wallyqs** [comment]
how big are the messages being published? is the client or subscription becoming a slow consumer possibly? You could try with `nats.PullMaxMessagesWithBytesLimit(10_000, 32*1024*1024)` to make sure that the messages being pulled stay below the default max pending size for the client and avoid disconnections.

**Comment by ljfuyuan** [comment]
Hi @wallyqs , Thanks for replying!
I just test again, the published message about 250 bytes and stream did not set limit.

The following are the status of stream and consumer when the problem occurs. I hope it will be helpful for analysis.

```
Information for Consumer TestStream > consumer-1 created 2025-08-02 15:22:29

Configuration:

                    Name: consumer-1
               Pull Mode: true
          Filter Subject: test.zone.1
          Deliver Policy: All
              Ack Policy: Explicit
                Ack Wait: 30.00s
           Replay Policy: Instant
      Maximum Deliveries: 3
         Max Ack Pending: 1,000
       Max Waiting Pulls: 512

Cluster Information:

                    Name: nats-cluster
                  Leader: nats-2

State:

            Host Version: 2.11.7
      Required API Level: 0 hosted at level 1
  Last Delivered Message: Consumer sequence: 119,096 Stream sequence: 119,096 Last delivery: 17.51s ago
    Acknowledgment Floor: Consumer sequence: 119,096 Stream sequence: 119,096 Last Ack: 17.50s ago
        Outstanding Acks: 0 out of maximum 1,000
    Redelivered Messages: 0
    ### Unprocessed Messages: 10
           Waiting Pulls: 0 of maximum 512

```

```
Information for Stream TestStream created 2025-08-02 15:22:29

                Subjects: test.zone.*
                Replicas: 1
                 Storage: File

Options:

               Retention: WorkQueue
         Acknowledgments: true
          Discard Policy: Old
        Duplicate Window: 2m0s
    Allows Batch Publish: false
         Allows Counters: false
       Allows Msg Delete: true
  Allows Per-Message TTL: false
            Allows Purge: true
          Allows Rollups: false

Limits:

        Maximum Messages: unlimited
     Maximum Per Subject: unlimited
           Maximum Bytes: unlimited
             Maximum Age: unlimited
    Maximum Message Size: unlimited
       Maximum Consumers: unlimited

Cluster Information:

                    Name: nats-cluster
                  Leader: nats-2

State:

            Host Version: 2.11.7
      Required API Level: 0 hosted at level 1
                Messages: 10
                   Bytes: 3.9 KiB
          First Sequence: 119,097 @ 2025-08-06 15:52:38
           Last Sequence: 119,106 @ 2025-08-06 15:52:38
        Active Consumers: 1
      Number of Subjects: 1
```
